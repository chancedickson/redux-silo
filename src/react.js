import React, { PureComponent, Children } from "react";

function createSiloProvider(store, Provider) {
  return class SiloProvider extends PureComponent {
    constructor(props) {
      super(props);

      this.state = { store, state: store.getState() };
    }
    componentDidMount() {
      store.subscribe(() => this.setState({ state: store.getState() }));
    }
    render() {
      return (
        <Provider value={this.state}>
          {Children.only(this.props.children)}
        </Provider>
      );
    }
  };
}

function createSiloConnector(Consumer) {
  return (reducerMap, mapStateToProps, mapDispatchToProps) => (Component) => {
    const componentName = Component.displayName || Component.name || null;
    const connectedName = componentName
      ? `SiloConnector<${componentName}>`
      : "SiloConnector";

    function connected() {
      return (
        <Consumer>
          {({ state, store }) => {
            const props = this.props;
            const passedState = reduceObject(
              reducerMap,
              (passedState, v, k) => {
                return {
                  ...passedState,
                  [k]: state.getIn(store.getReducerPath(v)),
                };
              },
              {},
            );
            const mappedState = mapStateToProps(passedState, props);

            return <Component {...props} {...mappedState} />;
          }}
        </Consumer>
      );
    }
    connected.displayName = connectedName;

    return connected;
  };
}

export function createSiloConnector(store) {
  const { Provider, Consumer } = React.createContext();

  return {
    Provider: createSiloProvider(store, Provider),
    connect: createSiloConnector(Consumer),
  };
}
