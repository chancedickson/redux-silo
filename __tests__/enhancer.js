import { Map, fromJS } from "immutable";
import { createStore } from "redux";
import { compileReducerTree, silo } from "../src/enhancer";

test("compileReducerTree.reducerByName matches snapshot.", () => {
  const reducerTree = {
    test1: () => {},
    test2: {
      test3: () => {},
    },
    test4: {
      test5: () => {},
      test6: () => {},
      test7: {
        test8: () => {},
        test9: {
          test10: () => {},
        },
      },
    },
  };
  const compiledReducerTree = compileReducerTree(reducerTree);
  expect(compiledReducerTree.reducerByName).toMatchSnapshot();
});

// TODO: Just test the reducer function and not createStore().
test("compileReducerTree.reducer calls all reducers and updates state properly.", () => {
  const store = createStore(
    {
      test1: (state = true, action) =>
        action.type === "FLIP" ? !state : state,
      test2: {
        test3: (state = 0, action) => {
          if (action.type === "INCREMENT") {
            return state + 1;
          } else if (action.type === "FLIP") {
            return -state;
          }
          return state;
        },
      },
    },
    silo,
  );

  expect(store.getState()).toEqual(
    fromJS({ test1: true, test2: { test3: 0 } }),
  );
  store.dispatch({ type: "INCREMENT" });
  expect(store.getState()).toEqual(
    fromJS({ test1: true, test2: { test3: 1 } }),
  );
  store.dispatch({ type: "FLIP" });
  expect(store.getState()).toEqual(
    fromJS({ test1: false, test2: { test3: -1 } }),
  );
});
