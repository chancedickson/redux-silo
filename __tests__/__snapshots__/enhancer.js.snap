// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileReducerTree.reducerByName matches snapshot. 1`] = `
Object {
  "test1": Array [
    "test1",
  ],
  "test10": Array [
    "test4",
    "test7",
    "test9",
    "test10",
  ],
  "test3": Array [
    "test2",
    "test3",
  ],
  "test5": Array [
    "test4",
    "test5",
  ],
  "test6": Array [
    "test4",
    "test6",
  ],
  "test8": Array [
    "test4",
    "test7",
    "test8",
  ],
}
`;
